chanbackup/single.go

```diff
@@ -0,0 +1,346 @@
+		)
+	}
+
+	// First we gather the SCB as is into a temporary buffer so we can
+	// determine the total length. Before we write out the serialized SCB,
+	// we write the length which allows us to skip any Singles that we
+	// don't know of when decoding a multi.
+	var singleBytes bytes.Buffer
+	if err := lnwire.WriteElements(
+		&singleBytes,
+		s.ChainHash[:],
+		s.FundingOutpoint,
+		s.ShortChannelID,
+		s.RemoteNodePub,
+		s.Addresses,
+		s.CsvDelay,
+		uint32(s.PaymentBasePoint.Family),
+		s.PaymentBasePoint.Index,
+		shaChainPub[:],
+		uint32(s.ShaChainRootDesc.KeyLocator.Family),
+		s.ShaChainRootDesc.KeyLocator.Index,
+	); err != nil {
+		return err
+	}
+
+	return lnwire.WriteElements(
+		w,
+		byte(s.Version),
+		uint16(len(singleBytes.Bytes())),
+		singleBytes.Bytes(),
+	)
+}
+
+// PackToWriter is similar to the Serialize method, but takes the operation a
+// step further by encryption the raw bytes of the static channel back up. For
+// encryption we use the chacah20poly1305 AEAD cipher with a 24 byte nonce and
+// 32-byte key size. We use a 24-byte nonce, as we can't ensure that we have a
+// global counter to use as a sequence number for nonces, and want to ensure
+// that we're able to decrypt these blobs without any additional context. We
+// derive the key that we use for encryption via a SHA2 operation of the with
+// the golden keychain.KeyFamilyStaticBackup base encryption key.  We then take
+// the serialized resulting shared secret point, and hash it using sha256 to
+// obtain the key that we'll use for encryption. When using the AEAD, we pass
+// the nonce as associated data such that we'll be able to package the two
+// together for storage. Before writing out the encrypted payload, we prepend
+// the nonce to the final blob.
+func (s *Single) PackToWriter(w io.Writer, keyRing keychain.KeyRing) error {
+	// First, we'll serialize the SCB (StaticChannelBackup) into a
+	// temporary buffer so we can store it in a temporary place before we
+	// go to encrypt the entire thing.
+	var rawBytes bytes.Buffer
+	if err := s.Serialize(&rawBytes); err != nil {
+		return err
+	}
+
+	// Finally, we'll encrypt the raw serialized SCB (using the nonce as
+	// associated data), and write out the ciphertext prepend with the
+	// nonce that we used to the passed io.Reader.
+	return encryptPayloadToWriter(rawBytes, w, keyRing)
+}
+
+// Deserialize attempts to read the raw plaintext serialized SCB from the
+// passed io.Reader. If the method is successful, then the target
+// StaticChannelBackup will be fully populated.
+func (s *Single) Deserialize(r io.Reader) error {
+	// First, we'll need to read the version of this single-back up so we
+	// can know how to unpack each of the SCB.
+	var version byte
+	err := lnwire.ReadElements(r, &version)
+	if err != nil {
+		return err
+	}
+
+	s.Version = SingleBackupVersion(version)
+
+	switch s.Version {
+	case DefaultSingleVersion:
+	default:
+		return fmt.Errorf("unable to de-serialize w/ unknown "+
```

Comment:
just making a note that once we add other versions, this should be deferred until after reading the length. then if we don't understand it, discard `length` bytes and return to the caller so that we advance the reader. preferably this would be a concrete error so that the multi reader can skip/log it during deserialization